### PR TITLE
Modernize build system to use pyproject.toml and build wheels

### PR DESCRIPTION
**Pull request by mayantaylor**
This PR aims to (1) modernize the build system to use pyproject.toml with the existing setup.py structure (see [python docs](https://packaging.python.org/en/latest/guides/modernize-setup-py-project/#modernize-setup-py-project)), and (2) set up Github Actions to build wheels automatically (using cibuildwheel) for distribution via Pypi.

Full instructions for creating and publishing wheels have been added to the wiki.

### CODE REVIEW

**Review by ritvikrao — APPROVED** [review]

**Comment by mayantaylor** [comment]
I'll be opening another PR to add tests to the binary distribution.

File: .github/workflows/build_wheels.yml [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+
+      - uses: actions/checkout@v4
+        with:
+          fetch-depth: 0 # Fetch full history, necessary for git describe in setup.py
+
+      - name: Checkout charm_src # is this best practice? should we use a submodule?
+        uses: actions/checkout@v4
+        with:
+          repository: charmplusplus/charm
+          token: ${{ secrets.GITHUB_TOKEN }}
```
Comment (matthiasdiener) [inline .github/workflows/build_wheels.yml:44]: Do we need the token here?

File: requirements.txt [hunk 2]
```diff
@@ -1,6 +1,3 @@
-cython>=3.0.0
-numpy
+numpy>=1.10.0
 greenlet>=3.0.0
-
-# Required for the charm++ build:
-cmake 
+cython>=3.0.0
```
Comment (matthiasdiener) [inline requirements.txt:3]: Just add a linebreak after the last line here

**Review by matthiasdiener — APPROVED** [review]
Besides my two minor comments, this looks good!

File: requirements.txt (hunk 2 above)
Comment (ritvikrao) [reply requirements.txt]: done

File: .github/workflows/build_wheels.yml (hunk 1 above)
Comment (mayantaylor) [reply .github/workflows/build_wheels.yml]: Looks like we don't, I removed it

**Review by ritvikrao — APPROVED** [review]

**Pull request merged**